### ISSUE TITLE
How to run playwright with bitbucket pipelines

### ISSUE THREAD

**Issue by MamdouhMansour** [microsoft/playwright-java#816]
I need to run my project with bitbucket pipeline and here is the error I get

##############
Host system is missing a few dependencies to run browsers. 
 Please install them with the following command:  
mvn exec:java -e -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install-deps"
<3 Playwright Team
############

**YML file:**

`image: maven:3.8-openjdk-17
#image: mcr.microsoft.com/playwright:focal

pipelines:
  branches:
    master: # Trigger this for any pushes to the master branch.
      - parallel:
          - step:
              script:
                # before script
                - mvn exec:java -e -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install webkit" -Dexec.args="install-deps"
                - mvn compiler:compile
          - step:
              # image: maven:3.8-openjdk-17
              name: Build and Test
              caches:
                - maven
              script:
                - mvn -B verify --file pom.xml
                - mvn install
              # - mvn -e test --no-transfer-progress --fail-at-end -Dtest="com.tests.gui.blinklos.TestLoanComplianceCheck, com.tests.gui.blinklos.TestLockedLEComplianceCheck,com.tests.gui.blinklos.TestBlinkLoanStagesAndDisclosures"
              after-script:
                # Collect checkstyle results, if any, and convert to Bitbucket Code Insights.
                - pipe: atlassian/checkstyle-report:0.3.0
          - step:
              name: Security Scan
              script:
                # Run a security scan for sensitive data.
                # See more security tools at https://bitbucket.org/product/features/pipelines/integrations?&category=security
                - pipe: atlassian/git-secrets-scan:0.5.1`

**Comment by yury-s** [comment]
> #image: mcr.microsoft.com/playwright:focal

Use java-specific docker [image](https://playwright.dev/java/docs/next/docker#pull-the-image), it will have all the browser dependencies for corresponding playwright version. If you want to use different base image you'll have to install those dependencies yourself, it can be done via [playwright cli](https://playwright.dev/java/docs/cli#install-system-dependencies). Bare in mind that only Ubuntu distribution is [officially supported](https://playwright.dev/java/docs/intro#linux).

**Comment by MamdouhMansour** [comment]
@yury-s 
Already updated yml file to java-specific docker, and still getting the same error


#  Template maven-build

#  This template allows you to test and build your Java project with Maven.
#  The workflow allows running tests, code checkstyle and security scans on the default branch.

# Prerequisites: pom.xml and appropriate project structure should exist in the repository.

image: mcr.microsoft.com/playwright:focal
pipelines:
  branches:
    master: # Trigger this for any pushes to the master branch.
      - step:
          services:
            - docker
          script:
            # before script
            - docker pull mcr.microsoft.com/playwright/java:v1.18.0-focal
            - docker run --rm mcr.microsoft.com/playwright/java:focal /bin/bash
      - step:
          image: atlassian/default-image:2
          name: Build and Test
          script:
            - mvn exec:java -e -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install-deps"
            - mvn install

          after-script:
            - pipe: atlassian/checkstyle-report:0.3.0

      - step:
          name: Security Scan
          script:
            - pipe: atlassian/git-secrets-scan:0.5.1
options:
  docker: true

**Comment by yury-s** [comment]
There was a regression in playwright: https://github.com/microsoft/playwright/issues/12075, it is fixed in v1.19.1 in nodejs playwright but the fix didn't make it to java yet.

**Comment by yury-s** [comment]
It shouldn't happen with 1.18 though. Can you reproduce the issue by running with docker command instead of bitbucket pipeline?

**Comment by MamdouhMansour** [comment]
will try this

**Comment by MamdouhMansour** [comment]
hi @yury-s  I tried and searched a lot, how can this be done?

**Comment by mxschmitt** [comment]
Your image does use `atlassian/default-image:2` and not the Playwright for Java Docker image. Thats probably the reason why it's not working.

**Comment by MamdouhMansour** [comment]
`Updated yml file to have 

image: mcr.microsoft.com/playwright:v1.18.0-focal
pipelines:
  branches:
    master: # Trigger this for any pushes to the master branch.
      - step:
          services:
            - docker
          script:
            # before script
            - docker pull mcr.microsoft.com/playwright/java:v1.18.0-focal
            - docker run --rm mcr.microsoft.com/playwright/java:v1.18.0-focal /bin/bash
            - docker build -t blink-testing:localbuild-focal -f Dockerfile.focal .
      - step:
          name: Build and Test
          script:
            - mvn -e test  --no-transfer-progress --fail-at-end -DtargetOperatingSystem="Linux-64" -DmaximumPerformanceMode="2" -Dtest="com.tests.gui.blinklos.TestLoanComplianceCheck, com.tests.gui.blinklos.TestLockedLEComplianceCheck,com.tests.gui.blinklos.TestBlinkLoanStagesAndDisclosures"

          after-script:
            # Collect checkstyle results, if any, and convert to Bitbucket Code Insights.
            - pipe: atlassian/checkstyle-report:0.3.0

      - step:
          name: Security Scan
          script:
            - pipe: atlassian/git-secrets-scan:0.5.1
options:
  docker: true`
  
  **Docker File**
  
  `FROM ubuntu:focal

# === INSTALL JDK and Maven ===

RUN apt-get update && apt-get install -y --no-install-recommends \
    openjdk-11-jdk maven

ENV JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64

# Install utilities required for downloading driver
RUN apt-get update && apt-get install -y --no-install-recommends \
    curl unzip

# === INSTALL playwright maven modules & browsers ===

# Browsers will remain downloaded in `/ms-playwright`.
# Note: make sure to set 777 to the registry so that any user can access
# registry.
ENV PLAYWRIGHT_BROWSERS_PATH=/ms-playwright

RUN mkdir /ms-playwright && chmod -R 777 $PLAYWRIGHT_BROWSERS_PATH

RUN mkdir /tmp/pw-java
COPY . /tmp/pw-java
RUN cd /tmp/pw-java && \
    mvn install -D skipTests --no-transfer-progress && \
    DEBIAN_FRONTEND=noninteractive mvn exec:java -e -D exec.mainClass=com.microsoft.playwright.CLI \
                     -D exec.args="install-deps" -f blink-testing/pom.xml --no-transfer-progress && \
    mvn exec:java -e -D exec.mainClass=com.microsoft.playwright.CLI \
                     -D exec.args="install" -f blink-testing/pom.xml --no-transfer-progress && \
    rm -rf /tmp/pw-java`
    
    
    and it gives me this-->
    
    Fatal error compiling: error: invalid target release: 17
    
    
![image](https://user-images.githubusercontent.com/39811691/155033793-6d4cdc8a-a75c-482d-bca1-965a7b6abfa5.png)

**Comment by mxschmitt** [comment]
You can remove these three lines, since the image is already pre built you don't need it:

```
mcr.microsoft.com/playwright/java:v1.18.0-focal
- docker run --rm mcr.microsoft.com/playwright/java:v1.18.0-focal /bin/bash
- docker build -t blink-testing:localbuild-focal -f Dockerfile.focal .
```

**Comment by MamdouhMansour** [comment]
**YML FIle:**
image: mcr.microsoft.com/playwright:v1.18.0-focal
pipelines:
  branches:
    master: # Trigger this for any pushes to the master branch.
      - step:
          services:
            - docker
          script:
            # before script
            - docker pull mcr.microsoft.com/playwright/java:v1.18.0-focal
            - docker run --rm mcr.microsoft.com/playwright/java:v1.18.0-focal /bin/bash
            - docker build -t blink-testing:localbuild-focal -f Dockerfile.focal .
      - step:
          name: Build and Test
          script:
#            - mvn exec:java -e -DtargetOperatingSystem="Linux-64" -Dexec.mainClass=com.microsoft.playwright.CLI -Dexec.args="install-deps"
            - mvn -e test  --no-transfer-progress --fail-at-end -DtargetOperatingSystem="Linux-64" -DmaximumPerformanceMode="2" -Dtest="com.tests.gui.blinklos.TestLoanComplianceCheck, com.tests.gui.blinklos.TestLockedLEComplianceCheck,com.tests.gui.blinklos.TestBlinkLoanStagesAndDisclosures"
          after-script:
            # Collect checkstyle results, if any, and convert to Bitbucket Code Insights.
            - pipe: atlassian/checkstyle-report:0.3.0

      - step:
          name: Security Scan
          script:
            # Run a security scan for sensitive data.
            # See more security tools at https://bitbucket.org/product/features/pipelines/integrations?&category=security
            - pipe: atlassian/git-secrets-scan:0.5.1
options:
  docker: true
  
  
  **DockerFile**
  FROM ubuntu:focal

# === INSTALL JDK and Maven ===

RUN apt-get update && apt-get install -y --no-install-recommends \
    openjdk-11-jdk maven

ENV JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64

# Install utilities required for downloading driver
RUN apt-get update && apt-get install -y --no-install-recommends \
    curl unzip

# === INSTALL playwright maven modules & browsers ===

# Browsers will remain downloaded in `/ms-playwright`.
# Note: make sure to set 777 to the registry so that any user can access
# registry.
ENV PLAYWRIGHT_BROWSERS_PATH=/ms-playwright

RUN mkdir /ms-playwright && chmod -R 777 $PLAYWRIGHT_BROWSERS_PATH

RUN mkdir /tmp/pw-java
COPY . /tmp/pw-java
RUN cd /tmp/pw-java && \
    mvn install -D skipTests --no-transfer-progress && \
    DEBIAN_FRONTEND=noninteractive mvn exec:java -e -D exec.mainClass=com.microsoft.playwright.CLI \
                     -D exec.args="install-deps" -f blink-testing/pom.xml --no-transfer-progress && \
    mvn exec:java -e -D exec.mainClass=com.microsoft.playwright.CLI \
                     -D exec.args="install" -f blink-testing/pom.xml --no-transfer-progress && \
    rm -rf /tmp/pw-java

**Comment by MamdouhMansour** [comment]
This didn't work, how can I run maven commands without setting up maven
![image](https://user-images.githubusercontent.com/39811691/155901648-795dcdce-7bdf-4699-bd34-40b26cef66fc.png)

**Comment by yury-s** [comment]
> Fatal error compiling: error: invalid target release: 17

Playwright docker image comes with [JDK 11](https://github.com/microsoft/playwright-java/blob/5ee8f23380353230b7de443c2dbf215806bafb08/Dockerfile.focal#L6), if you need a newer one you have to install it manually.

> This didn't work, how can I run maven commands without setting up maven

Which image is it? Playwright does have [Maven preinstalled](https://github.com/microsoft/playwright-java/blob/5ee8f23380353230b7de443c2dbf215806bafb08/Dockerfile.focal#L6).

**Comment by MamdouhMansour** [comment]
Here are the images

**Comment by yury-s** [comment]
> Here are the images 

In the second step configuration you override one image with another (I believe, but I'm not expert in bitbucket, refer to their docs). Also in the first step you use build docker image from docker file but then don't use it (and use the official image instead). In step one you user `docker ...` commands while in the second you call `mvn` directly, is that intentional?

**Comment by mxschmitt** [comment]
Looks answered and stale, closing by that. For further questions please file a new issue.
